ssg/entities/common.py

```diff
@@ -309,3 +349,85 @@ def update_with(self, rhs):
         updated_refinements = self._subtract_refinements(extended_refinements)
         updated_refinements.update(self.refine_rules)
         self.refine_rules = updated_refinements
+
+
+class Templatable(object):
+
+    KEYS = dict(
+        template=lambda: None,
+    )
+
+    def __init__(self):
+        pass
+
+    def is_templated(self):
+        return isinstance(self.template, dict)
+
+    def get_template_context(self, env_yaml):
+        # TODO: The first two variables, 'rule_id' and 'rule_title' are expected by some
+        #       templates and macros even if they are not rendered in a rule context.
+        #       Better name for these variables are 'entity_id' and 'entity_title'.
+        return {
+            "rule_id": self.id_,
+            "rule_title": self.title,
+            "products": env_yaml["product"],
+        }
+
+    def get_template_name(self):
+        """
+        Given a template dictionary from a Rule instance, determine the name
+        of the template (from templates) this rule uses.
+        """
+        try:
+            template_name = self.template["name"]
+        except KeyError:
+            raise ValueError(
+                "Templatable {0} is missing template name under template key".format(self))
+        return template_name
+
+    def get_template_vars(self, env_yaml):
+        if "vars" not in self.template:
+            raise ValueError(
+                "Templatable {0} does not contain mandatory 'vars:' key under "
+                "'template:' key.".format(self))
+        template_vars = self.template["vars"]
+
+        # Add the rule ID which will be used in template preprocessors (template.py)
+        # as a unique sub-element for a variety of composite IDs.
+        # TODO: The name _rule_id is a legacy from the era when rule was the only
+        #       context for a template. Preprocessors implicitly depend on this name.
+        #       A better name is '_entity_id' (as in XCCDF Entity).
+        template_vars["_rule_id"] = self.id_
```

Comment:
There is a good comment above it but I don't understand why a `_rule_id` is add here and at the same time a `rule_id` is add in `get_template_context`. Both items are set to `self.id_` . I have an impression that finally both of these variables will be passed to the template so it seems to be duplicate. Any thoughts?

Reply:
For some reason all `template.py` operate with `_rule_id` and `template.oval,ansible,etc` with marcos operate with `rule_id`.

Reply:
There seems to be no deliberate decision regarding the initial underscore. The variable has been introduced long ago (see https://github.com/ComplianceAsCode/content/blame/453a1d53dc0e2b129fc3704545264dd076e17889/ssg/templates.py#L359) and the leading underscore was probably supplied as a preemptive measure to prevent overwrites. I would drop the assignment here, and also resort to the underscore-less usage elsewhere. We use short IDs everywhere, there is nothing to fear.

Reply:
This dictionary element is used in:
- shared/templates/audit_rules_privileged_commands/template.py
- shared/templates/yamlfile_value/template.py
- shared/templates/file_groupowner/template.py
- shared/templates/file_owner/template.py
- shared/templates/file_permissions/template.py
- shared/templates/sysctl/template.py
- shared/templates/sebool/template.py
- ssg/utils.py

The `utils.py` parts are also being called from different `template.py`s. 

What is more important `templates.Teamplate.preprocess` is executed without extra environment (where `rule_id` and `rule_title` reside). On top of that `_rule_id` is transformed into `_RULE_ID` as the result of this process. It's a bit messy. So, we can't just drop it.

My suggestion is to leave it for another self-container round of refactoring.

Reply:
> My suggestion is to leave it for another self-contained round of refactoring.

I agree, please don't do this in this PR, we can get to that later, I have created a ticket to track this effort: https://github.com/ComplianceAsCode/content/issues/9882